File: 2019/4a.py
import re

inputStart = 372037
inputEnd = 905157

def solve():
  passwords = []

  for number in range(inputStart, inputEnd):
    # validate repeating adjacent digit
    isSixDigitsWithRepeating = re.match(r'(\d){0,6}\1', str(number))
    if not isSixDigitsWithRepeating:
      continue

    # validate decreasing/same digits
    decreases = False
    prevDigit = None
    for digit in map(int, str(number)):
      if prevDigit is None:
        prevDigit = digit
        continue
      else:
        if prevDigit > digit:
          decreases = True
          break
      prevDigit = digit

    if not decreases:
      passwords.append(number)
    
  return passwords

combinations = solve()
print(combinations)
print(f'valid passwords: {len(combinations)}')

File: 2019/4a.go
package main

import (
	"regexp"
	"strconv"
)

func main() {
	firstDigitConstant := 100
	inputStart := 372037
	inputEnd := 905157
	passwords := make([]int, 0)

	for number := inputStart; number <= inputEnd; number++ {
		// validate repeating adjacent digit
		isSixDigitsWithRepeating, _ := regexp.MatchString("(\\d){0,6}\\1", strconv.Itoa(number))
		if isSixDigitsWithRepeating {
			continue
		}

		// validate decreasing/same digits
		decreases := false
		prevDigit := firstDigitConstant
		for _, digit := range [...]int{number} {
			if prevDigit == firstDigitConstant {
				prevDigit = digit
				continue
			} else {
				if prevDigit > digit {
					decreases = true
					break
				}
			}
			prevDigit = digit
		}

		if !decreases {
			passwords = append(passwords, number)
		}
	}

	println("valid passwords: " + strconv.Itoa(len(passwords)))
}
